Initial Linux/LoongArch64 Port

xiangzhai opened the pull request:
#### 647e67b23883960fef8890465c0f70d7ab6e63f1
<pre>
Initial Linux/LoongArch64 Port
<a href="https://bugs.webkit.org/show_bug.cgi?id=244664">https://bugs.webkit.org/show_bug.cgi?id=244664</a>

Reviewed by Michael Catanzaro.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/WTF/wtf/PageBlock.h:
* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/dtoa/utils.h:
* Source/cmake/WebKitCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/254155@main">https://commits.webkit.org/254155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2db21334f8a39daab4cd9b7244e16be53931434f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88218 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/69/builds/32540 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97414 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152880 "Built successfully and passed tests") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/64/builds/30930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26744 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80380 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92081 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93827 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/64/builds/30930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74977 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24742 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/64/builds/30930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79812 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79967 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73721 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14758 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26168 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2914 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/31753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37670 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76565 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33926 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16991 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/JavaScriptCore/heap/HandleBlock.h:
```diff
@@ -39,7 +39,11 @@ class HandleBlock : public DoublyLinkedListNode<HandleBlock> {
+    static constexpr size_t blockSize = 16 * KB;
+#else
     static constexpr size_t blockSize = 4 * KB;
+#endif
```
Can you explain why this is needed, please? We already have MIPS support and it doesn't seem to need this. There's also no apparent relation to anything processor-specific?

mcatanzaro commented on Source/JavaScriptCore/heap/WeakBlock.h:
```diff
@@ -41,8 +41,11 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WeakBlock);
+#else
     static constexpr size_t blockSize = 256; // 1/16 of MarkedBlock size
-
+#endif
```
Same question here. (I see you found a stale comment, as 256 is definitely not guaranteed to be 1/16 of the MarkedBlock size. We should fix that, but the fix probably shouldn't be processor-specific like this.)

xiangzhai replied:
> We should fix that

I am familiar with OpenJDK hotspot JIT compilers and [LLVM](https://github.com/llvm/llvm-project/commits?author=xiangzhai).  But I am an apprentice in WebKit world :)  I will fix the "PAGE SIZE" issue,  due to LoongArch64 default page size is 16K, much more platform independent.

mcatanzaro commented on Source/cmake/WebKitCommon.cmake:
```diff
@@ -67,6 +67,8 @@ if (NOT HAS_RUN_WEBKIT_COMMON)
         set(COMPILER_IS_CLANG ON)
     endif ()
 
+    option(USE_LIBCXX "Use llvm libc++" OFF)
```
Mmm, I kinda suspect that most people who want libc++ will either pass the -stdlib=libc++ flag manually, or will get it by default, but I suppose it's harmless. It's surely not needed for porting to loongarch64, though, so I would split this out into a separate commit.

mcatanzaro replied:
Well actually, I see this is only needed if the platform default stdlib is libstdc++. So that means this option ensures that the compiled WebKit is incompatible with other system libraries. It's probably better not to provide this, right? It's an unusual thing to want to do, and you can always pass the flag manually if you really want it.

mcatanzaro replied:
Ah, I see you split this out into https://github.com/WebKit/WebKit/pull/3904.

mcatanzaro commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -196,6 +196,10 @@ if (LTO_MODE AND COMPILER_IS_CLANG)
 endif ()
 
+if (COMPILER_IS_CLANG AND USE_LIBCXX)
+    set(CMAKE_CXX_FLAGS "-std=c++20 -stdlib=libc++ ${CMAKE_CXX_FLAGS}")
```
You shouldn't mess with `-std=` here, though. That's not related to the chosen libc, and it's redundant with wherever else it gets set. Setting it here is asking for trouble next time we bump to a newer C++ version and forget to update it here.

mcatanzaro reviewed (changes requested):
This looks mostly uncontroversial, although I've requested a few changes. Except for one thing: you have some big unexplained changes in ArgumentCoders.h. Can you please explain in detail what you're doing there? It looks more likely to be related to trouble with a particular compiler version rather than CPU architecture?

xiangzhai committed:
Initial Linux/LoongArch64 Port
https://bugs.webkit.org/show_bug.cgi?id=244664

Reviewed by Michael Catanzaro.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/WTF/wtf/PageBlock.h:
* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/dtoa/utils.h:
* Source/cmake/WebKitCommon.cmake:

Canonical link: https://commits.webkit.org/254155@main
